### PR TITLE
Feature/holo 1480 lazy mint on init

### PR DESCRIPTION
**Pull request by Yashiru**
## Moved Lazy Mint into Init

The lazy mint feature needed to be more secure and prevent mistakes during deployment. This PR moves the lazy mint functionality to the initialization process.

### Changes

- The lazy mint feature has been moved to the `init` function and is no longer callable separately.
- Lazy mints are now executed using the `LazyMintConfiguration` struct in the `CustomERC721Initializer`.

> [!NOTE]
> Lazy mints are now handled by the deployment script `08_custom_erc721.ts`.

> [!WARNING]
> Some features of the `BatchMintMetadata` contract have been adapted for our custom use case.

### Testing

Run the related tests with the following command:
```bash
forge test --mc CustomERC721
```

Copy code

## Checklist before Requesting a Review

- [x] I have performed a self-review of my code.
- [x] Code styles have been enforced.
- [x] All Hardhat tests are passing.

### CODE REVIEW

File: src/token/CustomERC721.sol [hunk 1]
```diff
@@ -461,11 +498,12 @@ contract CustomERC721 is NonReentrant, ContractMetadata, InitializableLazyMint,
...
    *  for 'delayed reveal' NFTs.
+   *  TODO: Should we keep this function public ? (lazy mints are done in the init function)
    */
   function lazyMint(
     uint256 _amount,
-    string calldata _baseURIForTokens,
-    bytes calldata _data
+    string memory _baseURIForTokens,
+    bytes memory _data
   ) public override returns (uint256 batchId) {
```
Comment (alexanderattar) [inline src/token/CustomERC721.sol]: I think we should probably update it to be private?

File: scripts/utils/helpers.ts [hunk 2]
```diff
@@ -578,14 +578,14 @@ const txParams = async function ({
   let output: TransactionParamsOutput = {
     from: from as string,
     value: BigNumber.from(value),
-    gasLimit: gasLimit
-      ? '__gasLimitMultiplier' in global
-        ? gasLimit.mul(global.__gasLimitMultiplier).div(BigNumber.from('10000'))
-        : gasLimit
-      : await getGasLimit(hre, from as string, to as string, data, BigNumber.from(value), true),
+    // gasLimit: gasLimit
```
Comment (alexanderattar) [inline scripts/utils/helpers.ts]: Let's reenable this and remove the hardcoded one

File: src/extension/BatchMintMetadata.sol [hunk 3]
```diff
@@ -20,11 +20,11 @@ contract BatchMintMetadata {
   /// @dev Metadata frozen
   error BatchMintMetadataFrozen(uint256 batchId);
 
-  /// @dev Largest tokenId of each batch of tokens with the same baseURI + 1 {ex: batchId 100 at position 0 includes tokens 0-99}
-  uint256[] private batchIds;
+  /// @dev Largest tokenId of each batch of tokens with the same baseURIs + 1 {ex: batchId 100 at position 0 includes tokens 0-99}
+  uint256[] internal batchIds;
```
Comment (alexanderattar) [inline src/extension/BatchMintMetadata.sol:24]: what was this changed to `internal`?

File: src/extension/BatchMintMetadata.sol [hunk 4]
```diff
@@ -20,11 +20,11 @@ contract BatchMintMetadata {
...
 
   /// @dev Mapping from id of a batch of tokens => to base URI for the respective batch of tokens.
-  mapping(uint256 => string) private baseURI;
+  mapping(uint256 => string) internal baseURIs;
```
Comment (alexanderattar) [inline src/extension/BatchMintMetadata.sol:27]: why was this changed to plural?

File: src/extension/LazyMint.sol [hunk 5]
```diff
@@ -31,8 +31,8 @@ abstract contract LazyMint is ILazyMint, BatchMintMetadata {
    */
   function lazyMint(
     uint256 _amount,
-    string calldata _baseURIForTokens,
-    bytes calldata _data
+    string memory _baseURIForTokens,
```
Comment (alexanderattar) [inline src/extension/LazyMint.sol:34]: why the change from `calldata` to `memory`?

File: src/token/CustomERC721.sol [hunk 6]
```diff
@@ -162,16 +175,40 @@ contract CustomERC721 is NonReentrant, ContractMetadata, InitializableLazyMint,
...
 
   /**
-   * @notice Used internally to initialize the contract instead of through a constructor
-   * @dev This function is called by the deployer/factory when creating a contract
+   * @notice Initialize the lazy minting for the contract
+   * @dev This function also synchronizes the metadata with the prepended tokenID 
    */
   function initLazyMint() external override onlyOwner returns (uint256 chainPrepend) {
-    require(!_isLazyMintInitialized(), "HOLOGRAPH: lazy mint already initialized");
+    if(_isLazyMintInitialized()) revert LazyMint_AlreadyInitialized();
```
Comment (alexanderattar) [inline src/token/CustomERC721.sol]: prettier

File: test/foundry/fixtures/CustomERC721Fixture.t.sol [hunk 7]
```diff
@@ -200,9 +202,6 @@ contract CustomERC721Fixture is Test {
 
     // Connect the drop implementation to the drop proxy address
     customErc721 = CustomERC721(payable(newDropAddress));
-
-    vm.prank(DEFAULT_OWNER_ADDRESS);
-    return customErc721.initLazyMint();
```
Comment (alexanderattar) [inline test/foundry/fixtures/CustomERC721Fixture.t.sol:205]: should this be reenabled?

**Review by alexanderattar — CHANGES_REQUESTED** [review]
Some things look like they still need cleanup. Also can you remove the `yarn.lock` and start using `pnpm` going forward please.

File: src/extension/BatchMintMetadata.sol (hunk 3 above)
Comment (Yashiru) [reply src/extension/BatchMintMetadata.sol:24]: Changed this to allow the `initLazyMint` function to access it to synchronize the tokenId and the `nextTokenIdToLazyMint`. 

Check this function [here](https://github.com/holographxyz/holograph-protocol/blob/1e427a510412e23a3c5fb179ea301fc34e59e3be/src/token/CustomERC721.sol#L181)

File: src/extension/LazyMint.sol (hunk 5 above)
Comment (Yashiru) [reply src/extension/LazyMint.sol:34]: This function is only invoked within the `CustomERC721::lazyMint` function, which, in turn, is only invoked by `CustomERC721::init`. The latter holds `_baseURIForTokens` and `_data` within its in-memory parameter `initPayload`.

Given this context, both `_baseURIForTokens` and `_data` already reside in memory and cannot be accessed as calldata-located data in solidity.

We can also change the declaration of the `CustomERC721::init` from `function init(bytes memory initPayload)` to `function init(bytes calldata initPayload)`.

File: test/foundry/fixtures/CustomERC721Fixture.t.sol (hunk 7 above)
Comment (Yashiru) [reply test/foundry/fixtures/CustomERC721Fixture.t.sol:205]: The `initLazyMint` function don't do the actual lazyMints, it only sync the `CustomERC721` tokenId with the lazyMint one.

I have renamed it to `syncLazyMint` 👍🏻

File: src/token/CustomERC721.sol (hunk 6 above)
Comment (Yashiru) [reply src/token/CustomERC721.sol]: Running `pnpm run prettier:fix` doesn't change anything, I don't understand this one 😄

File: src/extension/LazyMint.sol (hunk 5 above)
Comment (alexanderattar) [reply src/extension/LazyMint.sol:34]: Okay cool thanks for the context @Yashiru !

File: src/token/CustomERC721.sol (hunk 6 above)
Comment (alexanderattar) [reply src/token/CustomERC721.sol]: it's missing space between `if(`

Comment (alexanderattar) [reply src/token/CustomERC721.sol]: I fixed it

File: src/token/CustomERC721.sol [hunk 8]
```diff
@@ -640,6 +662,26 @@ contract CustomERC721 is NonReentrant, ContractMetadata, InitializableLazyMint,
    * state changing
    */
 
+  /**
+   *  We override the `lazyMint` function, and use the `_data` parameter for storing encrypted metadata
+   *  for 'delayed reveal' NFTs.
+   *  TODO: Should we keep this function public ? (lazy mints are done in the init function)
```
Comment (alexanderattar) [inline src/token/CustomERC721.sol]: we can remove this comment

File: test/foundry/CustomERC721/CustomERC721.lazyMint.t.sol [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+contract CustomERC721LazyMintTest is CustomERC721Fixture, ICustomERC721Errors {
+  using Strings for uint256;
+
+  constructor() {}
+
+  function setUp() public override {
+    super.setUp();
+  }
+
+  function test_lazyMint() public setupTestCustomERC21(DEFAULT_MAX_SUPPLY) {
```
Comment (alexanderattar) [inline test/foundry/CustomERC721/CustomERC721.lazyMint.t.sol:22]: Is this supposed to be an empty test. Is it something you're still working on?
